src/modules/safe-shield/recipient-analysis/recipient-analysis.service.ts

```diff
@@ -1,7 +1,177 @@
+    private readonly erc20Decoder: Erc20Decoder,
+    @Inject(IConfigurationService)
+    private readonly configurationService: IConfigurationService,
+    @Inject(CacheService)
+    private readonly cacheService: ICacheService,
+    @Inject(LoggingService)
+    private readonly loggingService: ILoggingService,
+  ) {
+    this.defaultExpirationTimeInSeconds =
+      this.configurationService.getOrThrow<number>(
+        'expirationTimeInSeconds.default',
+      );
+  }
+  async analyze(args: {
+    chainId: string;
+    safeAddress: Address;
+    transactions: Array<DecodedTransactionData>;
+  }): Promise<RecipientAnalysisResponse> {
+    const recipients = extractRecipients(args.transactions, this.erc20Decoder);
+
+    const cacheDir = CacheRouter.getRecipientAnalysisCacheDir({
+      chainId: args.chainId,
+      recipients,
+    });
+
+    const cached = await this.cacheService.hGet(cacheDir);
+    if (cached) {
+      this.logCacheHit(cacheDir);
+      return JSON.parse(cached) as RecipientAnalysisResponse;
+    }
+    this.logCacheMiss(cacheDir);
+
+    const analysisResults: RecipientAnalysisResponse = {};
+    for (const recipient of recipients) {
+      const result = await this.analyzeRecipient({
+        chainId: args.chainId,
+        safeAddress: args.safeAddress,
+        recipient,
+      });
+
+      analysisResults[recipient] = result;
+    }
+
+    await this.cacheService.hSet(
+      cacheDir,
+      JSON.stringify(analysisResults),
+      this.defaultExpirationTimeInSeconds,
+    );
+
+    return analysisResults;
+  }
+
+  /**
+   * Analyzes the recipient and bridge status.
+   * @param args - The arguments for the analysis.
+   * @param args.chainId - The chain ID.
+   * @param args.safeAddress - The Safe address.
+   * @param args.recipient - The recipient address.
+   * @returns The analysis results.
+   */
+  async analyzeRecipient(args: {
+    chainId: string;
+    safeAddress: Address;
+    recipient: Address;
+  }): Promise<Record<RecipientStatusGroup, Array<RecipientAnalysisResult>>> {
+    const recipientInteractionResults = await this.analyzeInteractions(args);
+    return {
+      RECIPIENT_INTERACTION: [recipientInteractionResults],
+      BRIDGE: [],
+    };
+  }
+
+  /**
+   * Analyzes the interactions between a Safe and a recipient.
+   * @param args - The arguments for the analysis.
+   * @param args.chainId - The chain ID.
+   * @param args.safeAddress - The Safe address.
+   * @param args.recipient - The recipient address.
+   * @returns The analysis result.
+   */
+  async analyzeInteractions(args: {
+    chainId: string;
+    safeAddress: Address;
+    recipient: Address;
+  }): Promise<RecipientAnalysisResult> {
+    const transactionApi = await this.transactionApiManager.getApi(
+      args.chainId,
+    );
+
+    const page = await transactionApi.getTransfers({
+      safeAddress: args.safeAddress,
+      to: args.recipient,
+      limit: 1,
+    });
```

Comment:
It's clever that you're using a filtered tx-service request to get past transfers to a specific address! 🧠